cookiecutter/vcs.py

```diff
@@ -15,47 +18,268 @@
 logger = logging.getLogger(__name__)
 
 
-BRANCH_ERRORS = [
-    'error: pathspec',
-    'unknown revision',
+class VCS:
+    """Abstract base VCS class. Handles the cloning of VCS repositories."""
+
+    cmd = 'xyz'
+    """VCS command"""
+
+    allowed_protocols = ['http', 'https', 'ssh', 'file']
+    """
+    List of allowed protocols
+    (no match if protocol not among allowed or specific protocols)
+    """
+
+    identifiers = []
+    """VCS identifiers (implicit match if present in URL)"""
+
+    protocols = []
+    """VCS-specific protocols (e.g. git://, svn://)"""
+
+    suffixes = []
+    """VCS-specifix suffixes (e.g. .git, .hg)"""
+
+    not_found_errors = []
+    """
+    Error messages returned by the VCS if the repo was not found (lowercase)
+    """
+
+    branch_errors = []
+    """
+    Error messages returned by the VCS if the branch could not be checked out
+    (lowercase)
+    """
+
+    class MatchLevel(Enum):
+        """Match level returned by match_repo_url."""
+
+        NONE = 0
+        IMPLICIT = 1
+        EXPLICIT = 2
+
+    @classmethod
+    def match_repo_url(cls, url):
+        """
+        Check if the raw URL input from the matches this VCS.
+
+        Return match level:
+
+        | **NONE:** Does not match
+        | **IMPLICIT:** Contains VCS identifier
+          (e.g. https://private.com/svnrepo, gitolite@server:team/repo)
+        | **EXPLICIT:** Has VCS cmd prefix/protocol/suffix
+          (e.g. svn://private.com/svnrepo,
+          https://github.com/audreyr/cookiecutter.git,
+          hg+https://bitbucket.org/foo/bar)
+
+        :param url: Raw url parameter
+        :return: VCS.VCS.MatchLevel, repo_url (or None)
+        """
+        match = re.match(r'''(?:([\w\d]+)\+)?((\w+)://|\w+@)(\S+)''', url)
+
+        if not match:
+            return VCS.MatchLevel.NONE, ''
+
+        prefix, protocol_raw, protocol, repo_path = match.groups()
```

Comment:
This naming doesn't fit the case of ssh urls, check for example `git@gitorious.org:cookiecutter-gitorious/cookiecutter-gitorious.git`, the result would be:
```python
prefix, protocol_raw, protocol, repo_path = (
    None,
    'git@',
    None,
    'gitorious.org:cookiecutter-gitorious/cookiecutter-gitorious.git',
)
```

Even if everything works (`Git.match_repo_url` returns the expected result) it's only a coincidence and this could lead to bugs if someone changes the logic of `VCS.match_repo_url`.

What do you think about a separate regex for each protocol (file, http, ssh, vcs specific)? 
(I'm looking at https://git-scm.com/book/en/v2/Git-on-the-Server-The-Protocols while I write this, if I'm correct on the other VCSs it should work in the same way.)

It would allow us to write simpler regexes (less probability to be exposed to ReDoS) and more explicit code.

Reply:
Seems like a good idea. I'll look into rewriting that.

Another question: Cookiecutter matches Bitbucket urls to Hg automatically. As far as I know, Bitbucket switched from Hg to Git years ago. The current go-to for Hg hosting would be https://foss.heptapod.net. Shouldn't that be updated as well?

Reply:
Uh, good point, I was totally unaware of that.

Yes, the `elif 'bitbucket' in repo_url:` in `cookiecutter/vcs.py:42` is wrong nowadays. Hmm I don't know if it's a good idea to keep this behaviour and replace it with `foss.heptapod`, I think this pattern of matching a vcs depending on the hosting domain is just wrong (like bitbucket the underlying vcs could change over time and we can't add and mantain such logic for every different hosting solution).